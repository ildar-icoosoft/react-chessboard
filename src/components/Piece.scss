.piece {
  background-size: cover;
  width: 100%;
  height: 100%;
}

.bB {
  background-image: url(../images/pieces/black_bishop.svg);
}
.bK {
  background-image: url(../images/pieces/black_king.svg);
}
.bN {
  background-image: url(../images/pieces/black_knight.svg);
}
.bP {
  background-image: url(../images/pieces/black_pawn.svg);
}
.bQ {
  background-image: url(../images/pieces/black_queen.svg);
}
.bR {
  background-image: url(../images/pieces/black_rook.svg);
}

.wB {
  background-image: url(../images/pieces/white_bishop.svg);
}
.wK {
  background-image: url(../images/pieces/white_king.svg);
}
.wN {
  background-image: url(../images/pieces/white_knight.svg);
}
.wP {
  background-image: url(../images/pieces/white_pawn.svg);
}
.wQ {
  background-image: url(../images/pieces/white_queen.svg);
}
.wR {
  background-image: url(../images/pieces/white_rook.svg);
}
