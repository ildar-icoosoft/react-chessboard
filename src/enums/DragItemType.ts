export enum DragItemType {
  PIECE = "piece",
}
