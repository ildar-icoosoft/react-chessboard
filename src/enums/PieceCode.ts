export enum PieceCode {
  WHITE_PAWN = "wP",
  BLACK_PAWN = "bP",
  WHITE_KNIGHT = "wN",
  BLACK_KNIGHT = "bN",
  WHITE_BISHOP = "wB",
  BLACK_BISHOP = "bB",
  WHITE_ROOK = "wR",
  BLACK_ROOK = "bR",
  WHITE_QUEEN = "wQ",
  BLACK_QUEEN = "bQ",
  WHITE_KING = "wK",
  BLACK_KING = "bK",
}
