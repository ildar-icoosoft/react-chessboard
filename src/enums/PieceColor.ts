export enum PieceColor {
  WHITE = "white",
  BLACK = "black",
}
