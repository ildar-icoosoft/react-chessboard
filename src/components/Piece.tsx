import React, { CSSProperties, FC } from "react";
import { PieceCode } from "../enums/PieceCode";
import { getPieceElement } from "../utils/chess";
import { Transition } from "react-transition-group";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";
import { TransitionStatus } from "react-transition-group/Transition";
import { PartialRecord } from "../types/PartialRecord";
import {
  DEFAULT_SQUARE_WIDTH,
  DEFAULT_TRANSITION_DURATION,
} from "../constants/constants";

export interface PieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionFrom?: SquareTransitionFrom;
  transitionDuration?: number;
}

const getTransitionStyles = (
  transitionFrom: SquareTransitionFrom | undefined,
  state: TransitionStatus,
  transitionDuration: number
): CSSProperties => {
  if (!transitionFrom) {
    return {};
  }

  const styles: PartialRecord<TransitionStatus, CSSProperties> = {
    entering: {
      transform: `translate(${transitionFrom.x}px, ${transitionFrom.y}px)`,
    },
    entered: {
      transform: "translate(0, 0)",
      transition: `transform ${transitionDuration}ms`,
      zIndex: 1,
    },
  };

  return styles[state] || {};
};

export const Piece: FC<PieceProps> = ({
  pieceCode,
  width = DEFAULT_SQUARE_WIDTH,
  transitionFrom,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
}) => {
  return (
    <Transition appear={true} in={transitionFrom ? true : false} timeout={0}>
      {(state) => {
        return (
          <div
            className={"piece"}
            data-testid={`piece-${pieceCode}`}
            style={getTransitionStyles(
              transitionFrom,
              state,
              transitionDuration
            )}
          >
            <svg viewBox={`1 1 43 43`} width={width} height={width}>
              <g>{getPieceElement(pieceCode, { width: 45, height: 45 })}</g>
            </svg>
          </div>
        );
      }}
    </Transition>
  );
};
