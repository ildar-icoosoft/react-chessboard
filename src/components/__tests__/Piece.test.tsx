import React from "react";
import { Piece } from "../Piece";
import TestRenderer from "react-test-renderer";
import { PieceCode } from "../../enums/PieceCode";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

jest.useFakeTimers();

describe("Piece", () => {
  it("Snapshot", () => {
    const tree = TestRenderer.create(
      <Piece pieceCode={PieceCode.WHITE_KING} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("DOM structure", () => {
    it("contains data-testid piece-{pieceCode}", () => {
      const { queryByTestId, rerender } = render(
        <Piece pieceCode={PieceCode.WHITE_KING} />
      );
      expect(
        queryByTestId(`piece-${PieceCode.WHITE_KING}`)
      ).toBeInTheDocument();

      rerender(<Piece pieceCode={PieceCode.WHITE_QUEEN} />);
      expect(
        queryByTestId(`piece-${PieceCode.WHITE_QUEEN}`)
      ).toBeInTheDocument();
    });
  });
});
