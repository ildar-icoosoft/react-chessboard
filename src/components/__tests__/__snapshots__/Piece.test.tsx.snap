// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piece Snapshot 1`] = `
<div
  className="piece wK"
  data-testid="piece-wK"
/>
`;
